addons/mail/models/mail_link_preview.py

```diff
@@ -0,0 +1,160 @@
+        urls = tree.xpath('//a/@href')
+        link_previews = self.env['mail.link.preview']
+        for url in set(urls):
+            if (len(link_previews) >= 5):
+                break
+            link_previews |= self.env['mail.link.preview'].sudo()._create_link_preview(url, message.id)
+        if not link_previews:
+            return
+        guest = self.env['mail.guest']._get_guest_from_context()
+        if message.model == 'mail.channel' and message.res_id:
+            target = self.env['mail.channel'].browse(message.res_id)
+        elif self.env.user._is_public() and guest:
+            target = guest
+        else:
+            target = self.env.user.partner_id
+        self.env['bus.bus']._sendmany([(target, 'mail.link.preview/insert', link_previews._link_preview_format())])
+
+    @api.model
+    def _create_link_preview(self, url, message_id):
+        if self._is_domain_throttled(url):
+            return self.env['mail.link.preview']
+        link_preview_data = self._get_link_preview_from_url(url)
+        if link_preview_data:
+            link_preview_data['message_id'] = message_id
+            return self.sudo().create(link_preview_data)
+        return self.env['mail.link.preview']
+
+    def _delete_and_notify(self):
+        notifications = []
+        guest = self.env['mail.guest']._get_guest_from_context()
+        for link_preview in self:
+            if link_preview.message_id.model == 'mail.channel' and link_preview.message_id.res_id:
+                target = self.env['mail.channel'].browse(link_preview.message_id.res_id)
+            elif self.env.user._is_public() and guest:
+                target = guest
+            else:
+                target = self.env.user.partner_id
+            notifications.append((target, 'mail.link.preview/delete', {'id': link_preview.id}))
+        self.env['bus.bus']._sendmany(notifications)
+        self.unlink()
+
+    @api.model
+    def _is_link_preview_enabled(self):
+        link_preview_throttle = int(self.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99))
+        return link_preview_throttle > 0
+
+    @api.model
+    def _is_domain_throttled(self, url):
+        domain = urlparse(url).netloc
+        date_interval = fields.Datetime.to_string((datetime.now() - relativedelta(seconds=10)))
+        call_counter = self.search_count([
+            ('source_url', 'ilike', '%' + domain + '%'),
+            ('create_date', '>', date_interval),
+        ])
+        link_preview_throttle = int(self.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99))
+        return call_counter > link_preview_throttle
+
+    @api.model
+    def _get_link_preview_from_url(self, url):
+        try:
+            response = requests.head(url, timeout=3, headers=self.request_headers, allow_redirects=True)
+        except requests.exceptions.RequestException:
+            return False
+        if response.status_code != requests.codes.ok:
```

Comment:
Why not
```suggestion
        if not response.ok:
```
?

Reply:
Because `response.ok` covers a way too large range of status_code and we only want to allow `200 ok`.
From the request documentation (https://requests.readthedocs.io/en/latest/api/#requests.Response.ok):

```
This is not a check to see if the response code is 200 OK.
```